### ISSUE TITLE
AccessViolation when CommitAsync still in flight after Disposing

### ISSUE THREAD

**Issue by treziac** [confluentinc/confluent-kafka-dotnet#279]
Description
===========

CommitAsync can use librdkafka handle after consumer has been disposed, either provoking accessViolationException or thread blocked infinitely

How to reproduce
================

```
            using (var consumer = new Consumer<Null, string>(config, null, new StringDeserializer(Encoding.UTF8)))
            {
                consumer.Assign(new List<TopicPartitionOffset> { new TopicPartitionOffset("test", 0, Offset.Beginning) });
                consumer.CommitAsync(new[] { new TopicPartitionOffset("test", 0, 1)});
            }
```

Also, we may be stuck on `LibRdKafka.queue_poll(cQueue, -1/*infinite*/);` if handle was destroyed, and thread will never be exited. Maybe only wait for `message.timeout.ms` ? 
I'm not sure how to ensure the kafka handle is not disposed in all cases - a try catch will be the easier

Checklist
=========
Please provide the following information:

 - [ X] Confluent.Kafka nuget version: 0.11.0
 - [ X] Apache Kafka version: 0.11.0
 - [ ] Client configuration:
 - [X ] Operating system: windows 10
 - [ ] Provide logs (with "debug" : "..." as necessary in configuration)
 - [ ] Provide broker log excerpts
 - [X ] Critical issue

**Comment by treziac** [comment]
Personaly I don't like the CommitAsync being a Tak.Run.
I would rather make a CommitSync and let user do a Task.Run if he wants

**Comment by mhowlett** [comment]
thanks - well spotted. 
agree with you on `Task.Run`.

**Comment by roederja2** [comment]
I see an access violation when calling Dispose() on the consumer even if Commit is never called. In my case a Poll() is in progress on my polling thread. Is it not allowed to call Dispose() on a different thread than Poll() ?

**Comment by treziac** [comment]
You can call Dispose in any thread, but you should not call any other function on any other thread (poll or other) once you call dispoed. There is as possible race exception if you do this, resulting in AccessViolation.
As this is not related to this issue, if you have any problem please post in an other issue with all needed information and an excerpt of your code, thanks!

**Comment by roederja2** [comment]
Ok, so you are saying it is an issue in the library that should be fixed? Otherwise I will have to use a monitor to make sure Poll and Dispose never run at the same time I guess?

**Comment by treziac** [comment]
Yes, make sure you don't have any remaining call before disposing and after this.

This should be corrected on master via #262 (I though it did it on 0.11.0 but I messed up in targeting branch it seems), you can try with latest prerelease nupkg from appveyor. But I think when using multithreading, there can still be a rare race condition, which we can't correct without impacting performance.

**Comment by mhowlett** [comment]
in more recent versions, you'll get object disposed exceptions instead of access violation exceptions if you try to use the consumer after it's been disposed.

there is a related bug outstanding. if you don't manually dispose the consumer, when the finalizer is called by the GC, it will call any `OnPartitionAssigned` and `OnPartitionRevoked` handler you have registered - which will typically make use the `Consumer` instance. So the finalizer will result in an exception. this will be resolved in the 1.0 release.

**Comment by mhowlett** [comment]
for reference: #577
